Common mistakes in React (#1550)
enkidevs/curriculum: javascript/react/common-mistakes-in-react/setstate-is-async.md

---
author: kapnobatai136
  
aspects:

  - workout

  - deep

type: normal

category: tip

---

# setState is async

---
## Content

> More often than not, we are very eager with accessing a new state value after setting it. Because a new value is set on the next available render, the state might not reflect your latest update. Let's take a look at an example:

Reviewer: ```suggestion
More often than not, we are eager with accessing a new state value after setting it. Because a new value is set on the next available render, the state might not reflect your latest update. Let's take a look at an example:
```

```js
handleChange = age => {
  this.setState({ age });
  this.props.callback(this.state.age);
};
```

> In the `handleChange` function, when we call `this.props.callback(this.state.count)` we are getting the old state value, not the new one we set through `this.setState({ count })`. This issue can be fixed by using an optional second argument of the `setState` method, a callback function that is called after the state is updated:

Reviewer: ```suggestion
In the `handleChange` function, when we call `this.props.callback(this.state.count)` we are getting the old state value, not the new one we set through `this.setState({ count })` (because the state hasn't yet updated between the two lines of code). This issue can be fixed by using an optional second argument of the `setState` method, a callback function that is called after the state is updated:
```

```js
handleChange = age => {
  this.setState({ age }, () => {
    this.props.callback(this.state.age);
  });
};
```

If using functional components together with hooks, you would use:

```jsx
const [age, setAge] = useState(0);

useEffect(() => {
  callback(age);
}, [age, callback]);

const handleChange = value => {
  setAge(value);
};
```

> In the `useEffect()` hook, the `callback(age)` is called whenever the value of `age` changes. One important takeaway here is that `setState` is async, but not in a way that returns a promise. Using `async`, `await` or `then` methods will not work.

Reviewer: These two sentences aren't directly related so maybe split them up with a new line?

```suggestion
In the `useEffect()` hook, the `callback(age)` is called whenever the value of `age` changes. 

Note: Another important takeaway here is that although `setState` can be async, this happens behind the scenes, and we cannot control it. `setState` doesn't return a promise so using `await` or `then` will not work.
```
